sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusSenderAsyncClientRecoveryIsolatedTest.java

```diff
@@ -0,0 +1,1314 @@
+    private static final String QUEUE_NAME = "orders";
+    private static final String CLIENT_IDENTIFIER = "client-identifier";
+    private static final ServiceBusSenderInstrumentation DEFAULT_INSTRUMENTATION = new ServiceBusSenderInstrumentation(
+        null, null, FQDN, QUEUE_NAME);
+    private static final Duration VIRTUAL_TIME_SHIFT = OPERATION_TIMEOUT.plusSeconds(30);
+    private static final AmqpException RETRIABLE_LINK_ERROR = new AmqpException(true, AmqpErrorCondition.LINK_DETACH_FORCED,
+        "detach-link-error", new AmqpErrorContext(FQDN));
+    private static final AmqpException RETRIABLE_SESSION_ERROR = new AmqpException(true, "session-error",
+        new AmqpErrorContext(FQDN));
+    private static final AmqpException RETRIABLE_CONNECTION_ERROR = new AmqpException(true, AmqpErrorCondition.CONNECTION_FORCED,
+        "connection-forced-error", new AmqpErrorContext(FQDN));
+    private static final AmqpException NON_RETRIABLE_ERROR_1 = new AmqpException(false, AmqpErrorCondition.NOT_ALLOWED,
+        "not-allowed-error-1", new AmqpErrorContext(FQDN));
+    private static final AmqpException NON_RETRIABLE_ERROR_2 = new AmqpException(false, AmqpErrorCondition.NOT_ALLOWED,
+        "not-allowed-error-2", new AmqpErrorContext(FQDN));
+    private static final AmqpRetryOptions RETRY_OPTIONS = new AmqpRetryOptions()
+        .setMode(AmqpRetryMode.FIXED)
+        .setMaxRetries(10)
+        .setMaxDelay(Duration.ofSeconds(5))
+        .setDelay(Duration.ofSeconds(1))
+        .setTryTimeout(OPERATION_TIMEOUT);
+    private final MessageSerializer messageSerializer = new ServiceBusMessageSerializer();
+    @Mock
+    private Runnable onClientClosed;
+    @Captor
+    private ArgumentCaptor<List<Message>> sendMessagesCaptor0;
+    @Captor
+    private ArgumentCaptor<List<Message>> sendMessagesCaptor1;
+
+    private AutoCloseable mocksCloseable;
+
+    @BeforeEach
+    void setup() {
+        mocksCloseable = MockitoAnnotations.openMocks(this);
+    }
+
+    @AfterEach
+    void teardown() throws Exception {
+        Mockito.framework().clearInlineMock(this);
+        if (mocksCloseable != null) {
+            mocksCloseable.close();
+        }
+    }
+
+    @Test
+    @Execution(ExecutionMode.SAME_THREAD)
+    void shouldRecoverFromRetriableSendLinkError() {
+        // A Connection with one Session and two AmqpSendLink in that Session.
+        final int sessionsCnt = 1;
+        final int[] linksPerSession = new int[] { 2 };
+        try (MockEndpoint endpoint = createMockEndpoint(sessionsCnt, linksPerSession)) {
+            // The ServiceBusReactorAmqpConnection connection supplier for ReactorConnectionCache.
+            final Supplier<ServiceBusReactorAmqpConnection> connectionSupplier = singleConnectionSupplier(endpoint);
+            // Answers the invocation of 'send' API on two AmqpSendLink in the Session.
+            final Answer<Mono<Void>> sendAnswer = new Answer<Mono<Void>>() {
```

Comment:
I agree, it’s easier to read the lambda form and less lines. But the reason for choosing the anonymous form is – the function needs to use a counter which has to be outside the local scope of the function. I was trying to avoid declaring that counter as atomic or array with one element.